ngraph/core/reference/src/runtime/reference/fft.cpp

```diff
@@ -0,0 +1,533 @@
+                    }
+                }
+
+                bool blob_is_zero(const complex_type* data, int64_t blob_size)
+                {
+                    for (int64_t i = 0; i < blob_size; ++i)
+                    {
+                        if (data[i] != complex_type{0.0f, 0.0f})
+                        {
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+
+                int64_t offset_from_coords_and_strides(const std::vector<int64_t>& coords,
+                                                       const std::vector<int64_t>& strides)
+                {
+                    int64_t offset = 0;
+                    int64_t num_of_axes = coords.size();
+                    for (int64_t i = 0; i < num_of_axes; ++i)
+                    {
+                        offset += coords[i] * strides[i];
+                    }
+                    return offset;
+                }
+
+                void copy_data_to_output(complex_type* output,
+                                         const complex_type* data,
+                                         int64_t dst_index,
+                                         int64_t fft_size,
+                                         const std::vector<int64_t>& fft_strides,
+                                         const std::vector<int64_t>& output_fft_strides)
+                {
+                    for (int64_t idx = 0; idx < fft_size; ++idx)
+                    {
+                        auto coords = coords_from_index(idx, fft_strides);
+                        complex_type value = data[idx];
+                        int64_t offset = offset_from_coords_and_strides(coords, output_fft_strides);
+
+                        output[dst_index + offset] = value;
+                    }
+                }
+
+                static constexpr float pi = 3.141592653589793238462643f;
+
+                complex_type twiddle(int64_t k, int64_t length, FFTKind fft_kind)
+                {
+                    float angle = -2.0f * pi * static_cast<float>(k) / static_cast<float>(length);
+                    complex_type result = std::exp(complex_type(0.0f, angle));
+                    return (fft_kind == FFTKind::Inverse) ? std::conj(result) : result;
+                }
+
+                bool gather_to_buffer(const complex_type* data,
+                                      int64_t length,
+                                      int64_t start,
+                                      int64_t stride,
+                                      complex_type* buffer)
+                {
+                    bool input_is_zero = true;
+                    for (int64_t k = 0; k < length; ++k)
+                    {
+                        complex_type value = data[start + k * stride];
+                        buffer[k] = value;
+                        input_is_zero = input_is_zero && (value == complex_type(0.0f, 0.0f));
```

Comment:
We cannot leave the cycle if non zero element is found, because this function copies data to `buffer`. But we can rewrite this function with return type `void`, and use `blob_is_zero` to check that copied data are zeros.